events/pubsub.go

```diff
@@ -0,0 +1,159 @@
+package events
+
+import (
+	"fmt"
+	"github.com/spacemeshos/go-spacemesh/log"
+	"nanomsg.org/go-mangos"
+	"nanomsg.org/go-mangos/protocol/pub"
+	"nanomsg.org/go-mangos/protocol/sub"
+	"nanomsg.org/go-mangos/transport/ipc"
+	"nanomsg.org/go-mangos/transport/tcp"
+	"sync"
+	"time"
+)
+
+// channelBuffer defines the listening channel buffer size.
+const channelBuffer = 100
+
+// channelId is the Id on which subscribers must register in order to listen to messages on the channel.
+type channelId byte
+
+// Subscriber defines the struct of the receiving end of the pubsub messages.
+type Subscriber struct {
+	sock      mangos.Socket
+	output    map[channelId]chan []byte
+	allOutput chan []byte
+	chanLock  sync.RWMutex
+}
+
+// NewSubscriber received url string as input on which it will register to receive messages passed by server.
+func NewSubscriber(url string) (*Subscriber, error) {
+	socket, err := sub.NewSocket()
+	if err != nil {
+		return nil, err
+	}
+	socket.AddTransport(ipc.NewTransport())
+	socket.AddTransport(tcp.NewTransport())
+	err = socket.Dial(url)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Subscriber{
+		socket,
+		make(map[channelId]chan []byte),
+		nil,
+		sync.RWMutex{},
+	}, nil
+}
+
+// StartListening runs in a go routine and listens to all channels this subscriber is registered to. it then passes the
+// messages received to the appropriate reader channel.
+func (sub *Subscriber) StartListening() {
+	go func() {
+		for {
+			data, err := sub.sock.Recv()
+			log.Debug("got msg %v", string(data))
+			if err != nil {
+				if err.Error() == "connection closed" {
+					log.Warning("connection closed on pubsub reader")
+				} else {
+					log.Error("error on recv: %v", err)
+				}
+				return
+			}
+			// cannot route empty message
+			if len(data) < 1 {
+				log.Error("got empty message")
+			}
+			sub.chanLock.RLock()
+			if sub.allOutput != nil {
+				sub.allOutput <- data
+			}
+			if c, ok := sub.output[channelId(data[0])]; ok {
+				c <- data
+			}
+			sub.chanLock.RUnlock()
+		}
+	}()
+}
+
+// Close closes the socket which in turn will Close the listener func
+func (sub *Subscriber) Close() error {
+	return sub.sock.Close()
+}
+
+// Subscribe subscribes to the given topic, returns a channel on which data from the topic is received.
+func (sub *Subscriber) Subscribe(topic channelId) (chan []byte, error) {
+	if _, ok := sub.output[topic]; !ok {
```

Comment:
either move the lock above this line or first use a Read lock to check if this entry exists..